docs/src/main/asciidoc/amqp-reference.adoc

```diff
@@ -0,0 +1,647 @@
+
+@ApplicationScoped
+public class AmqpPriceConsumer {
+
+    @Incoming("prices")
+    public void consume(double price) {
+        // process your price.
+    }
+
+}
+----
+
+Or, you can retrieve the Message<Double>:
+
+[source, java]
+----
+package inbound;
+
+import org.eclipse.microprofile.reactive.messaging.Incoming;
+import org.eclipse.microprofile.reactive.messaging.Message;
+
+import javax.enterprise.context.ApplicationScoped;
+import java.util.concurrent.CompletionStage;
+
+@ApplicationScoped
+public class AmqpPriceMessageConsumer {
+
+    @Incoming("prices")
+    public CompletionStage<Void> consume(Message<Double> price) {
+        // process your price.
+
+        // Acknowledge the incoming message, marking the AMQP message as `accepted`.
+        return price.ack();
+    }
+
+}
+----
+
+=== Inbound Metadata
+Messages coming from AMQP contains an instance of `IncomingAmqpMetadata` in the metadata.
+
+[source, java]
+----
+Optional<IncomingAmqpMetadata> metadata = incoming.getMetadata(IncomingAmqpMetadata.class);
+metadata.ifPresent(meta -> {
+    String address = meta.getAddress();
+    String subject = meta.getSubject();
+    boolean durable = meta.isDurable();
+    // Use io.vertx.core.json.JsonObject
+    JsonObject properties = meta.getProperties();
+    // ...
+});
+----
+
+=== Deserialization
+
+The connector converts incoming AMQP Messages into Reactive Messaging `Message<T>` instances.
+`T` depends on the _body_ of the received AMQP Message.
+
+The http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html[AMQP Type System] defines the supported types.
+
+[options="header"]
+|===
+| AMQP Body Type | `<T>`
+| AMQP Value containing a http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#section-primitive-type-definitions[AMQP Primitive Type] | the corresponding Java type
+| AMQP Value using the `Binary` type | `byte[]`
+| AMQP Sequence | `List`
+| AMQP Data (with binary content) and the `content-type` is set to `application/json` | https://vertx.io/docs/apidocs/io/vertx/core/json/JsonObject.html[`JsonObject`]
+| AMQP Data with a different `content-type` | `byte[]`
+|===
+
+If you send objects with this AMQP connector (outbound connector), it gets encoded as JSON and sent as binary.
+The `content-type` is set to `application/json`.
+So, you can rebuild the object as follows:
+
+[source, java]
+----
+@ApplicationScoped
+public static class Consumer {
+
+    List<Price> prices = new CopyOnWriteArrayList<>();
+
+    @Incoming("from-amqp")
+    public void consume(JsonObject p) {   // <2>
+        Price price = p.mapTo(Price.class);  // <3>
+        prices.add(price);
+    }
+
+    public List<Price> list() {
+        return prices;
+    }
+}
+----
+1. The `Price` instances are automatically encoded to JSON by the connector
+2. You can receive it using a `JsonObject`
+3. Then, you can reconstruct the instance using the `mapTo` method
+
+NOTE: The `mapTo` method uses the Quarkus Jackson mapper.
```

Comment:
Maybe a link to how to configure it?

Reply:
And here is the fun part... we don't have documentation about that. It's somewhat covered in the rest-json guide.

Reply:
I've added a link.